File: leet_code/problems/easy/javascript/035.js
/*
Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.

You must write an algorithm with O(log n) runtime complexity.

Example 1:
Input: nums = [1,3,5,6], target = 5
Output: 2

Example 2:
Input: nums = [1,3,5,6], target = 2
Output: 1

Example 3:
Input: nums = [1,3,5,6], target = 7
Output: 4
*/

// Solution 1
const findPos1 = (input,target) => {
	for(let i = 0; i < input.length; i++) {
		if(input[i] > target || input[i] === target) {
			return i;
		}
		if(input[i] < target && input[i + 1] > target) {
		  return i + 1;
		}
		if(input[i] === input[input.length - 1]) {
		  return i + 1;
		}
	}
}

const input = [1,3,5,6];
const target = 9;
resp = findPos1(input, target);
console.log(resp1);


// Solution 2
function findPos2(nums, target) {
    let left = 0;
    let right = nums.length - 1;
    
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        
        if (nums[mid] === target) {
            return mid;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    // When the loop ends, left is the insertion point
    return left;
}
const input = [1,3,5,7,9,11,13,15,17,19];
const target = 4;
resp = findPos2(input, target);
console.log(resp1);

File: leet_code/problems/easy/python/035.py
"""
Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.

You must write an algorithm with O(log n) runtime complexity.

Example 1:
Input: nums = [1,3,5,6], target = 5
Output: 2

Example 2:
Input: nums = [1,3,5,6], target = 2
Output: 1

Example 3:
Input: nums = [1,3,5,6], target = 7
Output: 4
"""

from typing import List
import bisect


class Solution:
    def findPos1(self, nums: List[int], target: int) -> int:
        for i in range(len(nums)):
            if nums[i] >= target:
                return i
            if i + 1 < len(nums) and nums[i] < target < nums[i + 1]:
                return i + 1
        return len(nums)

    def findPos2(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return left

    def findPos3(self, nums: List[int], target: int) -> int:
        return bisect.bisect_left(nums, target)


nums = [1, 3, 5, 6]
target = 4
solObj = Solution()
resp = solObj.findPos1(nums, target)
print(resp)
